### PR TITLE
Adds chef plugin documentation

### PR DESCRIPTION
**Pull request by ares**

### CODE REVIEW

File: plugins/index.html [hunk 1]
```diff
@@ -32,6 +32,16 @@
                     <td><a href="http://www.katello.org/docs/">Documentation</a></td>
                 </tr>
             </table>
+            <table class="table">
+                <thead>
+                    <tr>
+                        <td><b>Foreman Chef</b></td>
```
Comment (domcleal) [inline plugins/index.html]: Nitpick, move this up so it's alphabetical

File: plugins/foreman_chef/0.1/index.md [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+infrastructure on dashboard, monitoring based on convergence report, 
+boostraping chef-client during host provisioning etc. This plugin also make
+active communication with chef server possible so we can delete nodes and
+clients upon host deletion in Foreman.
+
+All communication is through Foreman Proxy which is running a [separate plugin](https://github.com/theforeman/smart_proxy_chef) and has
+direct access to the Chef server and Foreman instance. Chef clients must be
+able to connect to Foreman Proxy while the opposite direction is not required. The upload is triggered at the end of each chef client run, which is managed by [third plugin](https://github.com/theforeman/chef-handler-foreman) in form of chef handler, also provided by Foreman.
+
+![Infrastructure overview](infrastructure_schema.svg)
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md]: You'll need to put the full path in here, as I think we have a `<base>` tag that makes it try and load from root.

File: plugins/foreman_chef/0.1/index.md [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+title: Foreman Chef 0.1 Manual
+version: 0.1
+---
+
+# 1. {{ page.title }}
+
+This plugin enables Foreman to receive reports and attributes from chef clients
+and use these data for usual Foreman features like displaying status of your
+infrastructure on dashboard, monitoring based on convergence report, 
+boostraping chef-client during host provisioning etc. This plugin also make
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md]: bootstrapping

File: plugins/foreman_chef/0.1/index.md [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+
+## 1.1 Supported platforms
+
+TODO (1.9 on smart proxy, foreman-tasks support)
+
+# 2. Installation
+
+There are two different components to install: Foreman Chef plugin and
+Foreman Proxy Chef plugin (aka Smart Proxy Chef). Since Foreman 1.8 the
+installer supports both and it's failry easy to setup them. Supposing
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md]: fairly

File: plugins/foreman_chef/0.1/index.md [hunk 5]
```diff
@@ -0,0 +1,474 @@
...
+you added Foreman repositories and installed foreman-installer package you
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md]: "installed the foreman-installer"

File: plugins/foreman_chef/0.1/index.md [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+can start Foreman installation using this command
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md]: "start the Foreman"

File: plugins/foreman_chef/0.1/index.md [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+
+    # to install Foreman with foreman_chef plugin
+    foreman-installer --enable-foreman-plugin-chef --enable-foreman-plugin-tasks
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md:53]: We ought to imply this somehow in the installer...

File: plugins/foreman_chef/0.1/index.md [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+
+At this moment, reports and attributes import should already work,
+supposing you configured you chef-clients correctly (see below). 
+Bootstrapping chef-client will need more simple manual steps. Also
+we'll go over available settings that you can use to modify Foreman
+Chef behaviour.
+
+## 3.1 Bootstrapping
+
+Bootstraping chef-client is natural part of host provisioning. If
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md]: Bootstrapping

File: plugins/foreman_chef/0.1/index.md [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+you're not familiar with how to setup Foreman provisioning, check
+out the manual - [provisioning section](http://www.theforeman.org/manuals/latest/index.html#4.4Provisioning). Foreman uses [provisioning templates](http://www.theforeman.org/manuals/latest/index.html#4.4.3ProvisioningTemplates), that allows
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md]: Remove the domain, just use /manuals/....

File: plugins/foreman_chef/0.1/index.md [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+and make all modifications to the clones. The reason is that newer version
+of Foreman Chef plugin can make changes to snippets.
+
+### 3.2 The actual bootstrap
+
+Now let's have a closer look to how these strategies work. Both gem and
+omnibus first installs a chef-client. The first strategy relies
+on existing system rubygems to install chef and chef_handler_foreman gems,
+to enable uploading reports and attributes to Foreman. The omnibus
+is using curl to download install.sh from https://www.chef.io/chef which
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md]: link?

File: plugins/foreman_chef/0.1/index.md [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+as there will be likely different version of kernel (e.g. some network
+interface names).
+
+### 3.3 Modify the behavior and select the strategy
+
+Now several question may pop up, e.g. how do we generate client.rb, how 
+client knows what's chef-server url, where is validation.pem read from,
+how do I choose which strategy to use, etc. The answer is similar to all
+of them. We modify the behaviour of provision templates using Foreman's
+parameters. You can read more about them in [Foreman manual](http://www.theforeman.org/manuals/latest/index.html#4.2.3Parameters), but the docs mention a lot about puppet,
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md]: ditto

File: plugins/foreman_chef/0.1/index.md [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+strategies in form of provisioning templates. These are only snippets
+that can be reused as a part of any provisioning template. We automatically
+include the bootstrap strategy in default `Kickstart default` and 
+`Preseed default` provision templates. Complete chef-client bootstrap
+strategy name is in format "chef-client $name bootstrap", currently you
+can find two of such name - **gem** and **omnibus**. There's a third
+snippet provided by the plugin called "chef-client bootstrap" which is
+a shared part for strategies mentioned before.
+
+![Bootstrap templates](templates.png)
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md]: ditto

File: plugins/foreman_chef/0.1/index.md [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+## 3.5 Manual chef-client setup
+
+If you don't want to use Foreman's provisioning capabilites and you
+prefer to install and configure your chef-clients manually or you want to
+register existing chef-nodes to Foreman, you can do it by your own.
+This way you still can use Foreman's dashboard, monitoring, reports etc. 
+
+All you have to do is to install chef_handler_foreman gem on every
+client and modify chef-client config file (/etc/chef/client.rb).
+ You have to put following lines at the end of you config files
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md]: "put the following", "your config files"

File: plugins/foreman_chef/0.1/index.md [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+attributes such as those set by recipes. One of attributes also
+contains the node runlist. On every following chef run the attributes
+are updated in Foreman. Foreman also parses these attributes to
+correctly assign Operating system, Environment etc.
+
+As you can see on following screenshow, we display all imported
+attributes. Since chef attributes are nested and forms tree structure,
+we display one level in a table.
+
+![Fact values list for one host](facts.png)
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md]: ditto on all of these

File: plugins/foreman_chef/0.1/index.md [hunk 15]
```diff
@@ -0,0 +1,474 @@
...
+on it, Foreman will generate a chart of distribution of this value throughout
+your infrastructure.
+
+![Value distribution](facts_distribution.png)
+
+To easily find all Hosts with chef-client of version you can click on that
+value. You'll be redirected to host list matching the search criteria. Note
+the search syntax above the table. You can search host based on any fact
+values. The same syntax works everywhere in Foreman. For more details
+about the syntax, you can check the [scoped_search project](https://github.com/wvanbergen/scoped_search) that we use.
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md]: better to link to section 4.1.5 of the main manual

File: plugins/foreman_chef/0.1/index.md [hunk 16]
```diff
@@ -0,0 +1,474 @@
...
+key for this node and uses it to verify the signature. 
+If everything is OK, it forward the request to Foreman.
+See a diagram below for illustration.
+
+![Authentication](authentication.svg)
+
+## 5.2 Future
+
+Plan
+* create a client in chef-server during provisioning, so no validation.pem is needed
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md:458]: Add a line break between "Plan" and this line so it's formatted correctly.

File: plugins/foreman_chef/0.1/index.md [hunk 17]
```diff
@@ -0,0 +1,474 @@
...
+
+
+## 7.2 Contributing
+
+Follow the [same process as Foreman]({{site.baseurl}}contribute.html#SubmitPatches)
+for contributing.
+
+# 8. Links
+
+* foreman_chef plugin https://github.com/theforeman/foreman_chef
```
Comment (domcleal) [inline plugins/foreman_chef/0.1/index.md]: link these?

**Comment by domcleal** [comment]
Looks good, thanks!

File: plugins/foreman_chef/0.1/index.md (hunk 7 above)
Comment (ares) [inline plugins/foreman_chef/0.1/index.md:53]: Completely agree, ideas welcome, I hoped to have scenarios support in kafo already but it does not seem to be worked on anytime soon.

**Comment by ares** [comment]
@domcleal I think I addressed all your comments and all paths starts with `/manual` or `/plugins`. I also appreciate your :gb: review

File: plugins/foreman_chef/0.1/index.md [hunk 18]
```diff
@@ -0,0 +1,475 @@
...
+
+![Infrastructure overview](/plugins/foreman_chef/0.1/infrastructure_schema.svg)
+
+Foreman Chef does support "legacy" mode where communication is direct
+with no Foreman Proxy involved. In this mode we do not handle authentication
+of any uploads from Chef clients, therefore it's strongly discouraged.
+
+## 1.1 Supported platforms
+
+TODO (1.9 on smart proxy, foreman-tasks support)
```
Comment (mmoll) [inline plugins/foreman_chef/0.1/index.md]: Could we have just a short note on 1.8 and 1.9?

File: plugins/foreman_chef/0.1/index.md (hunk 7 above)
Comment (stbenjam) [inline plugins/foreman_chef/0.1/index.md:53]: Does the installer do something that pulling the foreman-tasks package via dependencies won't?

**Comment by ares** [comment]
@mmoll I filled `1.1 Supported platforms` chapter with some actual info, thanks

**Comment by mmoll** [comment]
:+1: - I'm not sure if @stbenjam's comment got lost here, but I also wonder why there's foreman-tasks in the installer line, when the package already depends on it... :panda_face:

**Comment by domcleal** [comment]
Maybe so the service is properly enabled, which the class does.

**Comment by ares** [comment]
I replied to him through email because the comment was not here anymore... Installer enables and start the foreman-tasks service. That's why it's needed.

**Comment by mmoll** [comment]
I'd expect the Package also to enable and start the service, but OK... merging :)

**Comment by mmoll** [comment]
thanks @ares :fireworks:

**Comment by ares** [comment]
@mmoll AFAIK that's not recommended for RPMs, pretty standard in DEBs :-) thanks guys, I'm happy to announce manual tomorrow

**Comment by domcleal** [comment]
We could just put an "includes foreman::plugin::tasks" into the chef class, come to think of it.

**Comment by ares** [comment]
if foreman tasks provide some configuration options they wouldn't be exposed, I think we can leave it as is until we have scenarios support in the installer

**Pull request merged**